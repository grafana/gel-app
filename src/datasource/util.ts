import { TempGELQueryWrapper } from './types';
import { DataFrame, DataFrameHelper, dateTime, FieldType } from '@grafana/data';

export function getNextQueryID(query: TempGELQueryWrapper) {
  if (!query || !query.queries) {
    return 'GA';
  }
  const A = 'A'.charCodeAt(0);
  const ids = query.queries.map(q => q.refId);
  for (let i = query.queries.length; i < 27; i++) {
    const id = 'G' + String.fromCharCode(A + i);
    if (!ids.includes(id)) {
      return id;
    }
  }
  return 'G' + Date.now(); //
}

export function gelResponseToDataFrames(rsp: any[]): DataFrame[] {
  return rsp.map((v: any) => {
    const frame = new DataFrameHelper();
    frame.name = v.name;
    frame.refId = v.refId;
    if (v.labels) {
      frame.labels = v.labels;
    }
    for (const f of v.fields) {
      let v = f.values;
      const type: FieldType = f.type;
      // HACK: this should be supported out-of-the-box
      if (type === FieldType.time) {
        v = v.map((str: string) => {
          return dateTime(str).valueOf();
        });
      }
      frame.addField({
        ...f,
        values: v,
      });
    }
    return frame;
  });
}
