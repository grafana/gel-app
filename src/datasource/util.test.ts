import { responseToDataFrame } from './util';

const resp = {
  Values: [
    {
      Name: 'AAA',
      Fields: [
        {
          Name: 'Time',
          Type: 4,
          Unit: '',
          Vector: [
            '2019-08-20T10:29:00-04:00',
            '2019-08-20T10:30:00-04:00',
            '2019-08-20T10:31:00-04:00',
            '2019-08-20T10:32:00-04:00',
            '2019-08-20T10:33:00-04:00',
            '2019-08-20T10:34:00-04:00',
          ],
        },
        { Name: '', Type: 1, Unit: '', Vector: [2, 4, 4, 4, 2, 2] },
      ],
      Labels: null,
    },
  ],
};

describe('PluginDatasource', () => {
  describe('when querying', () => {
    test('should return the saved data with a query', () => {
      const frame = responseToDataFrame(resp)[0];
      expect(frame.name).toEqual('AAA');
      expect(frame.fields.length).toEqual(2);
      expect(frame.fields[0].name).toEqual('Time');
      expect(frame.length).toEqual(resp.Values[0].Fields[0].Vector.length);
    });
  });
});
