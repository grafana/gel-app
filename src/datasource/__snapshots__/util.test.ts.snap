// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GEL Utils should parse sample GEL output 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "name": "Time",
        "type": "time",
        "values": Array [
          1566937080000,
        ],
      },
      Object {
        "config": Object {},
        "name": "GE-series",
        "type": "number",
        "values": Array [
          2,
        ],
      },
    ],
    "labels": undefined,
    "meta": undefined,
    "name": "AAA",
    "refId": "GE",
  },
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "name": "Time",
        "type": "time",
        "values": Array [
          1566937114433,
          1566937115433,
          1566937116433,
          1566937117433,
          1566937118433,
        ],
      },
      Object {
        "config": Object {},
        "name": "GA-series",
        "type": "number",
        "values": Array [
          60.466028797961954,
          60.906537886006966,
          61.07109793922545,
          61.00881212641243,
          60.933449623483696,
        ],
      },
    ],
    "labels": undefined,
    "meta": undefined,
    "name": "BBB",
    "refId": "GA",
  },
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "name": "Field 1",
        "type": "number",
        "values": Array [
          60.8771852746181,
        ],
      },
    ],
    "labels": undefined,
    "meta": undefined,
    "name": "CCC",
    "refId": "GC",
  },
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "name": "Time",
        "type": "time",
        "values": Array [
          1566937114433,
          1566937115433,
          1566937116433,
          1566937117433,
          1566937118433,
        ],
      },
      Object {
        "config": Object {},
        "name": "Field 2",
        "type": "number",
        "values": Array [
          0.9932461319490806,
          1.0004821611126804,
          1.0031853093031915,
          1.0021621704617347,
          1.0009242271733128,
        ],
      },
    ],
    "labels": undefined,
    "meta": undefined,
    "name": "DDD",
    "refId": "GD",
  },
]
`;
